File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/SecuredSessionsViewModel.kt
// "Open secured sessions" manager (docs/service-access-gating.md,
// "Web-experience gating"). Lists the browser QR-login sessions this phone
// authorized (from SecuredSessionStore), lets the user Refresh a row's
// online/offline (rate-limited ~1/min/secretId by the box — debounced ≥60s
// client-side too; a 429 keeps the last-known state) and Stop a session (closes
// it on the box, then drops it from the store).

package com.flagshipserver.app.viewmodels

import androidx.lifecycle.ViewModel
import com.flagshipserver.app.api.ServiceAccessClient
import com.flagshipserver.app.core.HttpException
import com.flagshipserver.app.core.SecuredSession
import com.flagshipserver.app.core.SecuredSessionStore
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Per-session liveness as the user has most recently observed it. */
enum class SessionLiveness { UNKNOWN, ONLINE, OFFLINE }

data class SecuredSessionRow(
    val session: SecuredSession,
    val liveness: SessionLiveness = SessionLiveness.UNKNOWN,
    /** Last time we successfully queried (ms); drives the ≥60s debounce. */
    val lastCheckedAt: Long? = null,
    val refreshing: Boolean = false,
)

class SecuredSessionsViewModel(
    private val client: ServiceAccessClient = ServiceAccessClient(),
    private val now: () -> Long = { System.currentTimeMillis() },
    private val load: () -> List<SecuredSession> = { SecuredSessionStore.list() },
    private val removeFromStore: (String) -> Unit = { SecuredSessionStore.remove(it) },
) : ViewModel() {
    /** Client-side status debounce; mirror of the box's ~1/min rate limit. */
    private val refreshDebounceMs = 60_000L

    private val _rows = MutableStateFlow<List<SecuredSessionRow>>(emptyList())
    val rows: StateFlow<List<SecuredSessionRow>> = _rows.asStateFlow()

    /** (Re)load the persisted sessions, preserving any liveness already observed. */
    fun reload() {
        val prior = _rows.value.associateBy { it.session.secretId }
        _rows.value = load().map { s ->
            prior[s.secretId]?.copy(session = s) ?: SecuredSessionRow(session = s)
        }
    }

    /** True when a refresh would be rate-limited (debounce not yet elapsed). */
    fun canRefresh(secretId: String): Boolean {
        val row = _rows.value.firstOrNull { it.session.secretId == secretId } ?: return false
        val last = row.lastCheckedAt ?: return true
        return now() - last >= refreshDebounceMs
    }

    suspend fun refresh(secretId: String) {
        val row = _rows.value.firstOrNull { it.session.secretId == secretId } ?: return
        if (row.refreshing) return
        // Client-side debounce: a 429 from the box otherwise just wastes a call.
        if (!canRefresh(secretId)) return
        update(secretId) { it.copy(refreshing = true) }
        try {
            val status = client.sessionStatus(row.session.serverId, secretId)
            update(secretId) {
                it.copy(
                    liveness = if (status == "online") SessionLiveness.ONLINE else SessionLiveness.OFFLINE,
                    lastCheckedAt = now(),
                    refreshing = false,
                )
            }
        } catch (e: HttpException) {
            // 429 (rate limited) or any transport error: keep the last-known
            // liveness, just clear the spinner. Mark lastCheckedAt only on a
            // genuine 429 so the debounce backs the user off.
            update(secretId) {
                it.copy(
                    refreshing = false,
                    lastCheckedAt = if (e.status == 429) now() else it.lastCheckedAt,
                )
            }
        } catch (e: Throwable) {
            update(secretId) { it.copy(refreshing = false) }
        }
    }

    /** Stop a session: close it on the box (best-effort), then drop it locally. */
    suspend fun stop(secretId: String) {
        val row = _rows.value.firstOrNull { it.session.secretId == secretId } ?: return
        runCatching { client.closeSession(row.session.serverId, secretId) }
        removeFromStore(secretId)
        _rows.value = _rows.value.filterNot { it.session.secretId == secretId }
    }

    private fun update(secretId: String, transform: (SecuredSessionRow) -> SecuredSessionRow) {
        _rows.value = _rows.value.map { if (it.session.secretId == secretId) transform(it) else it }
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/SecuredSessionsViewModel.swift
import Foundation
import Observation
import FlagshipAPI
import FlagshipCore

/// Settings → "Open secured sessions" (docs/service-access-gating.md,
/// "Web-experience gating"). Lists the browser QR-login sessions THIS phone has
/// authorized (held locally in `SecuredSessionStoring`), lets the owner refresh
/// each one's `online`/`offline` (rate-limited ~1/min/secretId), and stop one
/// (close on the box + remove locally).
///
/// The box rate-limits status to ~1/min/secretId (429 = "checked too
/// recently"). We debounce client-side too: a Refresh within `minRefreshMs`
/// (default 60s) of the last is a no-op that surfaces a subtle "checked
/// recently" hint rather than hitting the box. A server 429 (the two clocks can
/// disagree) is handled identically — we keep the last known status.
@Observable
@MainActor
public final class SecuredSessionsViewModel {

    public private(set) var sessions: [SecuredSession] = []
    /// Last-known liveness per secretId (nil = not yet checked this launch).
    public private(set) var statuses: [String: SecuredSessionStatus] = [:]
    /// Secrets currently mid-refresh — drives the per-row spinner.
    public private(set) var refreshing: Set<String> = []
    /// Secrets whose last Refresh was debounced/429'd — drives the subtle
    /// "checked recently" hint. Cleared the moment a real refresh lands.
    public private(set) var recentlyChecked: Set<String> = []

    private let client: any ServiceAccessClient
    private let store: any SecuredSessionStoring
    /// Per-secretId last-real-query wall time (ms) for the client debounce.
    private var lastQueryAt: [String: Int64] = [:]
    private let minRefreshMs: Int64
    private let now: () -> Int64

    public init(
        client: any ServiceAccessClient,
        store: any SecuredSessionStoring,
        minRefreshMs: Int64 = 60_000,
        now: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970 * 1000) }
    ) {
        self.client = client
        self.store = store
        self.minRefreshMs = minRefreshMs
        self.now = now
    }

    /// (Re)load the held sessions from the local store.
    public func load() {
        sessions = store.list()
        // Drop status/debounce state for sessions that no longer exist.
        let live = Set(sessions.map { $0.secretId })
        statuses = statuses.filter { live.contains($0.key) }
        lastQueryAt = lastQueryAt.filter { live.contains($0.key) }
        recentlyChecked = recentlyChecked.intersection(live)
    }

    /// Whether a Refresh on `session` would be debounced (too soon since the
    /// last real query). The row can grey its Refresh accordingly.
    public func canRefresh(_ session: SecuredSession) -> Bool {
        guard let last = lastQueryAt[session.secretId] else { return true }
        return now() - last >= minRefreshMs
    }

    /// Refresh a single session's status. Debounced ≥`minRefreshMs` per
    /// secretId; a server 429 keeps the last status + flags "checked recently".
    public func refresh(_ session: SecuredSession) async {
        let id = session.secretId
        if refreshing.contains(id) { return }
        if let last = lastQueryAt[id], now() - last < minRefreshMs {
            recentlyChecked.insert(id)
            return
        }
        refreshing.insert(id)
        defer { refreshing.remove(id) }
        do {
            let status = try await client.sessionStatus(serverDomain: session.serverId, secretId: id)
            lastQueryAt[id] = now()
            statuses[id] = status
            recentlyChecked.remove(id)
        } catch ServiceAccessError.statusRateLimited {
            // The box's clock said too-soon; keep the last status + hint.
            lastQueryAt[id] = now()
            recentlyChecked.insert(id)
        } catch {
            // Transient — leave the last-known status untouched.
        }
    }

    /// Stop a session: close it on the box (idempotent), then drop it locally.
    /// The local removal proceeds even if the close call fails (the session
    /// still expires on its own on the box; the owner wanted it gone here).
    public func stop(_ session: SecuredSession) async {
        let id = session.secretId
        do {
            try await client.closeSession(serverDomain: session.serverId, secretId: id)
        } catch {
            // best-effort — remove locally regardless.
        }
        store.remove(secretId: id)
        statuses.removeValue(forKey: id)
        lastQueryAt.removeValue(forKey: id)
        recentlyChecked.remove(id)
        sessions.removeAll { $0.secretId == id }
    }
}
